File: packaging/mcpb/kotlin/stackql-mcp/src/main/kotlin/io/stackql/mcp/Cache.kt
package io.stackql.mcp

import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.security.MessageDigest

/**
 * The shared on-disk binary cache, `~/.stackql/mcp-server-bin`. Same location
 * the npm and pypi wrappers use, so multiple embedders on one machine share a
 * single extraction, and so do polyglot toolchains.
 *
 * Layout: `<root>/<version>/<platform-key>/stackql[.exe]`.
 */
object Cache {
    /** Env override: path to a stackql binary to run directly, skipping acquisition. */
    const val ENV_BIN: String = "STACKQL_MCP_BIN"

    /** Env override: path to a local `.mcpb` to extract instead of downloading. */
    const val ENV_BUNDLE: String = "STACKQL_MCP_BUNDLE"

    /** `~/.stackql/mcp-server-bin`. */
    fun defaultRoot(): Path =
        Paths.get(System.getProperty("user.home"), ".stackql", "mcp-server-bin")

    /** The extraction directory for a version/platform: `<root>/<version>/<key>`. */
    fun binaryDir(root: Path, version: String, platform: Platform): Path =
        root.resolve(version).resolve(platform.key)

    /** The canonical binary path inside [binaryDir]. */
    fun binaryPath(root: Path, version: String, platform: Platform): Path =
        binaryDir(root, version, platform).resolve(platform.exeName)

    /** Lowercase hex sha256 of a byte array. */
    fun sha256(data: ByteArray): String =
        MessageDigest.getInstance("SHA-256").digest(data).toHex()

    /** Lowercase hex sha256 of a file, streamed so large binaries do not load whole. */
    fun sha256(path: Path): String {
        val digest = MessageDigest.getInstance("SHA-256")
        Files.newInputStream(path).use { input ->
            val buf = ByteArray(64 * 1024)
            while (true) {
                val n = input.read(buf)
                if (n < 0) break
                digest.update(buf, 0, n)
            }
        }
        return digest.digest().toHex()
    }

    private fun ByteArray.toHex(): String =
        joinToString("") { "%02x".format(it) }
}

File: packaging/mcpb/cargo/src/cache.rs
//! Cache layout and environment overrides.
//!
//! The binary cache is shared with the npm and PyPI wrappers:
//! `~/.stackql/mcp-server-bin/<version>/<platform-key>/` - always check it
//! before downloading.

use std::path::PathBuf;

use crate::error::{Error, Result};

/// Path to a stackql binary to run directly, skipping acquisition entirely.
pub const ENV_BIN: &str = "STACKQL_MCP_BIN";
/// Path to a local .mcpb bundle to extract instead of downloading.
pub const ENV_BUNDLE: &str = "STACKQL_MCP_BUNDLE";

/// Resolve the user's home directory without external crates.
pub fn home_dir() -> Result<PathBuf> {
    if let Some(home) = std::env::var_os("HOME").filter(|v| !v.is_empty()) {
        return Ok(PathBuf::from(home));
    }
    if cfg!(windows) {
        if let Some(profile) = std::env::var_os("USERPROFILE").filter(|v| !v.is_empty()) {
            return Ok(PathBuf::from(profile));
        }
    }
    Err(Error::NoHomeDir)
}

/// Default approot: `<home>/.stackql`.
pub fn default_approot() -> Result<PathBuf> {
    Ok(home_dir()?.join(".stackql"))
}

/// Root of the shared binary cache: `<home>/.stackql/mcp-server-bin`.
pub fn bin_cache_root() -> Result<PathBuf> {
    Ok(default_approot()?.join("mcp-server-bin"))
}

/// Cache directory for one extracted bundle:
/// `<home>/.stackql/mcp-server-bin/<version>/<platform-key>/`.
pub fn bundle_cache_dir(version: &str, platform_key: &str) -> Result<PathBuf> {
    Ok(bin_cache_root()?.join(version).join(platform_key))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cache_dir_matches_the_shared_layout() {
        let dir = bundle_cache_dir("0.10.500", "linux-x64").unwrap();
        let suffix: PathBuf = [".stackql", "mcp-server-bin", "0.10.500", "linux-x64"]
            .iter()
            .collect();
        assert!(
            dir.ends_with(&suffix),
            "{} should end with {}",
            dir.display(),
            suffix.display()
        );
    }
}
